src/BaGet.Core/Mirror/MirrorService.cs

```diff
@@ -32,53 +32,42 @@ public class MirrorService : IMirrorService
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
         }
 
-        public static MirrorService Create(
-            IPackageService localPackages,
-            NuGetClient client,
-            IPackageIndexingService indexer,
-            ILogger<MirrorService> logger)
-        {
-            return new MirrorService(
-                localPackages,
-                new MirrorV3Client(client),
-                indexer,
-                logger);
-        }
-
-        public async Task<IReadOnlyList<NuGetVersion>> FindPackageVersionsOrNullAsync(
+        public async Task<IReadOnlyList<NuGetVersion>> FindPackageVersionsAsync(
             string id,
             CancellationToken cancellationToken)
         {
-            var upstreamVersions = await RunOrNull(
-                id,
-                "versions",
-                () => _upstreamClient.ListPackageVersionsAsync(id, includeUnlisted: true, cancellationToken));
-
-            if (upstreamVersions == null || !upstreamVersions.Any())
-            {
-                return null;
-            }
+            var upstreamVersions = await _upstreamClient.ListPackageVersionsAsync(id, cancellationToken);
 
             // Merge the local package versions into the upstream package versions.
             var localPackages = await _localPackages.FindAsync(id, includeUnlisted: true, cancellationToken);
             var localVersions = localPackages.Select(p => p.Version);
 
+            if (!upstreamVersions.Any()) return localVersions.ToList();
+            if (!localPackages.Any()) return upstreamVersions;
+
             return upstreamVersions.Concat(localVersions).Distinct().ToList();
         }
 
-        public async Task<IReadOnlyList<Package>> FindPackagesOrNullAsync(string id, CancellationToken cancellationToken)
+        public async Task<IReadOnlyList<Package>> FindPackagesAsync(string id, CancellationToken cancellationToken)
         {
-            var items = await RunOrNull(
-                id,
-                "metadata",
-                () => _upstreamClient.GetPackageMetadataAsync(id, cancellationToken));
+            var upstreamPackageMetadata = await _upstreamClient.GetPackageMetadataAsync(id, cancellationToken);
+
+            var upstreamPackages = upstreamPackageMetadata.Select(ToPackage).ToList();
+            var localPackages = await _localPackages.FindAsync(id, includeUnlisted: true, cancellationToken);
+
+            if (!upstreamPackages.Any()) return localPackages;
+            if (!localPackages.Any()) return upstreamPackages;
 
-            if (items == null || !items.Any())
+            // Merge the local packages into the upstream packages.
+            var result = upstreamPackages.ToDictionary(p => new PackageIdentity(p.Id, p.Version));
+            var local = localPackages.ToDictionary(p => new PackageIdentity(p.Id, p.Version));
```

Comment:
TODO:
1. Test local package does not exist
2. Test upstream package does not exist
3. Test local and upstream does not exist
4. Test local and upstream exists